File: daily/C++版本/141. 环形链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        // if(!head)    //哈希表法
        //     return false;
        // ListNode *p=head;
        // unordered_set<ListNode*> s;
        // while(s.find(p)==s.end())
        // {
        //     s.insert(p);
        //     p=p->next;
        //     if(!p)
        //         return false;
        // }
        // return true;
        
        if(!head)   //快慢指针法
            return false;
        ListNode *p=head,*q=head;
        while(p && p->next)
        {
            p=p->next->next;
            q=q->next;
            if(p==q)
                return true;
        }
        return false;  
    }
};	

File: daily/双指针/双指针easy/141. 环形链表.js

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    //哈希表法
    // if(!head)
    //     return false;
    // var p=head;
    // var map=new Map();
    // map.set(p,true);
    // while(p.next)
    // {
    //     if(map.has(p.next))
    //         return true;
    //     map.set(p.next,true);
    //     p=p.next;
    // }
    // return false;

    //快慢指针法
    // if(!head)
    //     return false;
    // var p=head,q=head;
    // while(p && p.next)
    // {
    //     p=p.next.next;
    //     q=q.next;
    //     if(p==q)
    //         return true;
    // }
    // return false;

    //骚做法，JSON.stringify不能字符串化含循环引用的结构
    // try{
    //     JSON.stringify(head)
    //         return false;
    // }catch(err){
    //     return true;
    // }
};